Cursor Overlay Bullet Positioning

linghong opened the pull request:
Ticket #1175 

**1. replace the triangle and the dot with svg**
The main reason of the triangle/ dot  didn't place on the center of the overlay circle is because the shapes were made by  text.  When aligning text to the center vertically, it align the whole text line. If you ever looked at an individual letter after a sentence is vertically centered, you would see the letter doesn't always stay on the center vertically. For example for the "." text, the text "." occupies a rectangle space instead of a square, and the '.' doesn't stay on the center of the rectangle in all the  browsers.

Previously, to make them centered, we use margin or padding to try to make them looked centered, but they won't work correctly in different font size and browsers.

**2. moved the cursor overlay to the bullet: using 'border-radius: 50%' with a grey background color, we can achieve the circle overlay**, 
This ease the burden of adjusting the positioning of  original Bullet-Cursor-Overlay Circle to align with bullet.

I adjusted the CSS code to make it also centered after clicking the top-tight icons, and also not breaking any existing features/ looks due to this move.  

**3. I haven't adjusted the .android and .mobile CSS code yet** 
In the end the JS code related to the Bullet-Cursor-Overlay code and all the CSS code related to the Bullet-Cursor-Overlay should be removed to simplify the code base, but I would like to submit the code to get it reviewed to first ensure I didn't break up any existing features.

shresthabijay commented:
I have noticed few problems.

## Drag hover style has broken

**Broken**
<img width="459" alt="Screen Shot 2021-08-16 at 14 14 35" src="https://user-images.githubusercontent.com/25879100/129534753-f74c4ba3-742d-4c23-a2d3-5ca2406aa8d1.png">

**Expected**

<img width="299" alt="Screen Shot 2021-08-16 at 13 56 54" src="https://user-images.githubusercontent.com/25879100/129533255-bc779a0c-090e-42bb-9a27-17230ad6011c.png">

##   Thought annotation position

You can see the `ThoughtAnnotation` `2` is not positioned correctly.

**Broken**
<img width="588" alt="Screen Shot 2021-08-16 at 14 00 45" src="https://user-images.githubusercontent.com/25879100/129534866-cf2e998b-3a4d-4c19-a4a7-2a0d37a9ca79.png">

**Expected**
<img width="492" alt="Screen Shot 2021-08-16 at 14 01 08" src="https://user-images.githubusercontent.com/25879100/129534890-da448759-424e-4760-abfd-808b426ae1ec.png">


## Full editable should be clickable.

Only a minimum part of `Editable` clickable. `Editable` should expand the container. In the video, you can notice that any click a little bit right to the editable is being registered outside, instead of focusing on editable.

**Broken**
https://user-images.githubusercontent.com/25879100/129535732-f6ba1bea-f11d-4084-8ddb-b8b338d2bf8f.mov

**Expected**
https://user-images.githubusercontent.com/25879100/129535798-6ac007d5-ec96-4fca-8736-db8a06d96942.mov

shresthabijay commented:
@linghong I have pushed some changes that does the following:

- Makes bullet icons size change according to global `fontSize`
- Fixes few problems with bullet not maintaining its width when editable is of multiple rows.
- Removed transform origin from the bullet.

shresthabijay commented:
@linghong  any updates ?

shresthabijay commented:
> @shresthabijay
> When I tested the code on android, it asks me to provide an invite code.

@linghong I am assuming you are accessing the web application using local IP. For testing, you can change the value of `isLocalNetwork` here to `true`.

https://github.com/cybersemics/em/blob/21a733edb8fa99d0afe7c2231407512bbe4a5e42/src/components/ModalWelcome.tsx#L14

linghong commented:
@shresthabijay  I have done with the fixation, but I have a few things unclear:
1. when I ran the current dev branch, I got an error as the below, what things I need to set up on my computer to make it work?
> Unhandled Rejection (OpenFailedError): NotFoundError Failed to execute 'transaction' on 'IDBDatabase': One of the specified object stores was not found.
 NotFoundError: Failed to execute 'transaction' on 'IDBDatabase': One of the specified object stores was not found.

2. this branch has conflicts with the dev already, Is it correct to solve the conflicts as the below:  git rebase it to the dev branch, go through all the commits to fix the conflicts, then forcefully push it to the remote branch?

Thanks!

raineorshine commented:
> 1. when I ran the current dev branch, I got an error as the below, what things I need to set up on my computer to make it work?
> 
> > Unhandled Rejection (OpenFailedError): NotFoundError Failed to execute 'transaction' on 'IDBDatabase': One of the specified object stores was not found.
> > NotFoundError: Failed to execute 'transaction' on 'IDBDatabase': One of the specified object stores was not found.

The IndexedDB schema was updated. I posted a message on Discord on the #dev channel, but you may have missed it. You can fix it by fully clearing your site data:

<img width="372" alt="Screen Shot 2021-09-11 at 3 55 11 PM" src="https://user-images.githubusercontent.com/750276/132960008-40f7834e-344d-4e81-979f-48f8708a9fd5.png">


> 1. this branch has conflicts with the dev already, Is it correct to solve the conflicts as the below:  git rebase it to the dev branch, go through all the commits to fix the conflicts, then forcefully push it to the remote branch?

Yes, exactly!

linghong commented:
@raineorshine  Thank you!

linghong commented:
@raineorshine  there is a new error message on the dev branch
Clean the site data, go to the dev site, click "this ain't my first redo" to enter it, enter a few text, then reload the site, the site crashes with the below error: 

> Unhandled Rejection (TypeError): Cannot read properties of undefined (reading 'lastUpdated')
(anonymous function)
src/action-creators/loadLocalState.ts:11
   8 | /** Loads the local state from the IndexedDB database. */
   9 | const loadLocalState = (): Thunk<Promise<void>> => async dispatch => {
  10 |   // load helpers and settings from local database
> 11 |   const [{ lastUpdated, recentlyEdited }, settings] = await Promise.all([
  12 |     db.getHelpers() as Promise<db.Helper>,
  13 |     getContext(db, [EM_TOKEN, 'Settings']),
  14 |   ])

raineorshine commented:
> @raineorshine there is a new error message on the dev branch
> Clean the site data, go to the dev site, click "this ain't my first redo" to enter it, enter a few text, then reload the site, the site crashes with the below error:
> 
> > Unhandled Rejection (TypeError): Cannot read properties of undefined (reading 'lastUpdated')
> > (anonymous function)
> > src/action-creators/loadLocalState.ts:11
> > 8 | /** Loads the local state from the IndexedDB database. */
> > 9 | const loadLocalState = (): Thunk<Promise> => async dispatch => {
> > 10 |   // load helpers and settings from local database
> > 11 |   const [{ lastUpdated, recentlyEdited }, settings] = await Promise.all([
> > 12 |     db.getHelpers() as Promise<db.Helper>,
> > 13 |     getContext(db, [EM_TOKEN, 'Settings']),
> > 14 |   ])

I'm getting that too. There must have been a regression. I will track it down to the offending commit.

FYI It would be better to post problems with `dev` to the #dev channel on Discord. They are unrelated to this issue as such.

linghong commented:
@shresthabijay @raineorshine: could you take a look at my commits when you have time? Thanks!

shresthabijay commented:
@linghong Good work! Everything is working. Just the empty child drop was a little off to the right.

<img width="530" alt="Screen Shot 2021-09-17 at 14 36 14" src="https://user-images.githubusercontent.com/25879100/133755119-3e4958c6-c871-46c6-b24e-f9f40dc6e35c.png">

It was a little change, so I fixed it myself.

shresthabijay commented:
@raineorshine Can you make sure if everything is working for you  ?

raineorshine commented:
I'm still seeing issues in solving the original problem.

At Font Size 17 on iOS Safari you can see that the triangle is too far to the right. Also the triangle is too big and the bullet is too small. The size of the bullets should not change from `dev`, just the alignment.

![IMG_0148](https://user-images.githubusercontent.com/750276/133778764-985107b3-f13f-4ee6-b96e-b397098f4f34.jpg)

At Font Size 13 on iOS Safari you can see that the triangle is too far to the left:

![IMG_0147](https://user-images.githubusercontent.com/750276/133778763-179d6be5-317e-41a5-a6f8-69132e676828.jpg)

Note that I researched different approaches and came to the conclusion that separate elements would always have subpixel deviations: https://github.com/cybersemics/em/issues/1175#issuecomment-877716865. I suggested a unified SVG for the bullet and cursor overlay. I hope that you can demonstrate that your approach is viable before going further down this path.

linghong commented:
@raineorshine: Do you want the bullet, triangle and dot size stay same? I think @shresthabijay want their size responsive to the font size change.  


> Note that I researched different approaches and came to the conclusion that separate elements would always have subpixel deviations: [#1175 (comment)](https://github.com/cybersemics/em/issues/1175#issuecomment-877716865). I suggested a unified SVG for the bullet and cursor overlay. I hope that you can demonstrate that your approach is viable before going further down this path.

I'll check what cause the problem

raineorshine commented:
> @raineorshine: Do you want the bullet, triangle and dot size stay same? I think @shresthabijay want their size responsive to the font size change.

We want the bullet, triangle and dot size to vary with the font size so that the relative proportion to the text remains constant, just like in `dev`. The size should be the same as currently in `dev`.

shresthabijay commented:
@raineorshine Previously I had misunderstood some things about this task. Now I realize that. Sorry for that. However, I think the solution you presented seems to work. Here is a simple demonstration in a codesandbox.

[Bullet overlay with unified SVG example](https://m6cr9.csb.app)

[Related codesandbox](https://codesandbox.io/s/bullet-with-overlay-m6cr9?file=/index.html)

@raineorshine Can you check if there are misalignments in the ios safari? I tested it in firefox, chrome, macOS safari, and ios safari (simulator). It seems to work. If you confirm that it is viable, @linghong can work on this taking the example as reference.

raineorshine commented:
The demonstration works correctly at all sizes on iOS Safari.

shresthabijay commented:
@linghong Please implement the bullet cursor overlay taking the reference from the above-mentioned sandbox example.  Let me know if you need help with anything.

raineorshine commented:
The misalignments in the PR are also present on desktop Chrome, even at non-extreme font sizes. However, I do think the situation has improved.

Here is a perfectly centered triangle at Font Size 18 as of the latest commit in this PR (85e9a7de85):

![image](https://user-images.githubusercontent.com/750276/134021276-b0411a2e-dc21-4a91-bfab-71fef5bbea27.png)

When the Font Size is increased to 19, the triangle is too far to the left:

![image](https://user-images.githubusercontent.com/750276/134021350-4088b325-8101-4146-a3b9-4b44f3374a45.png)

This effect becomes more dramatic at smaller font sizes. The discrepancy occurs at odd font sizes due to subpixel rounding. This is why I believe that two separate elements will always yields subpixel deviations at small font sizes. A single SVG will not produce such deviations since it is scaled down uniformly.

Let's start by fixing the alignment across font sizes (say 9px to 29px) on desktop Chrome only. Do not test on other platforms until this is consistent.

Then we will test on mobile Safari and other platforms.

linghong commented:
@raineorshine @shresthabijay Please test whether it is working on your machines. Thanks! 
@raineorshine Since you want me to try the current approach first.  I think the only way can do is  increasing/decreasing 2px each time. If that approach doesn't work out on your machines, we have to switch to replace the overlay circle with SVG graph.

raineorshine commented on src/components/Bullet.tsx:
```diff
@@ -52,10 +60,26 @@ const Bullet = ({
-      {glyph || (showContexts ? (isLeaf ? '◦' : '▹') : isLeaf ? '•' : '▸')}
+    <span
+      className='glyph'
+      style={{ width: `${Math.round(fontSize / 2) * 2}px`, height: `${Math.round(fontSize / 2) * 2}px` }}
```
Let's add a descriptive comment explaining the need for the rounding since it is not obvious from the code. You can include a link to the issue.

raineorshine reviewed (changes requested):
> @raineorshine @shresthabijay Please test whether it is working on your machines. Thanks!

It looks great! Thank you!

> @raineorshine
> Since you want me to try the current approach first. I think the only way can do is letting the overly/dot/triangle size only increasing or decreasing when it reaches every 2px of changes with each 1px of the font size goes up or down.

Since the irregularities only occur at odd font sizes, this covers it!

> Also when I told to you that I couldn't see the dot/triangle was off-center on my computer if I increased or decreased the font size, I actually tested it by changing the fontSize on the Bullet.tsx file. -I didn't know it has a font size +/- button on the bottom.

No problem. It would be the same effect either way.

So the last items before we merge are:

- [ ] Fix triangle proportions (size and shape) to match `dev`.
- [x] Rebase on `dev` and fix merge conflicts
- [x] Fix table view (see below)

raineorshine commented:
The gray circle incorrectly renders for all table grandchildren rather than only when the cursor is on the thought.

## Steps to Reproduce

```js
- a
  - =view
    - Table
  - b
    - c
  - d
    - e
```

## Current Behavior

With the cursor set to `null`:

![image](https://user-images.githubusercontent.com/750276/134773014-c2d11678-044e-441e-bf58-5fa0ae3c1bc3.png)

## Expected Behavior

![image](https://user-images.githubusercontent.com/750276/134773048-cd8a72b8-680f-42e5-b229-6cd0150e36f8.png)

linghong commented:
@raineorshine  Please review the code. Thank you!
1. Fixed all the things you mentioned
2. Fixed another issue: the color of the triangles/dots for the thought  that the cursor  is pointing at and for its child thoughts are supposed to be a bright white color, but was a faint white color.

raineorshine commented:
Closing in favor of #1505.
